rotkehlchen/exchanges/kucoin.py

```diff
@@ -0,0 +1,775 @@
+import base64
+import hashlib
+import hmac
+import logging
+from collections import defaultdict
+from datetime import datetime
+from enum import Enum
+from functools import partial
+from http import HTTPStatus
+from json.decoder import JSONDecodeError
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    overload,
+)
+from urllib.parse import urlencode
+
+import gevent
+import requests
+from gevent.lock import Semaphore
+from requests.adapters import Response
+from typing_extensions import Literal
+
+from rotkehlchen.accounting.structures import Balance
+from rotkehlchen.assets.asset import Asset
+from rotkehlchen.assets.converters import asset_from_kucoin
+from rotkehlchen.constants.misc import ZERO
+from rotkehlchen.errors import (
+    DeserializationError,
+    RemoteError,
+    SystemClockNotSyncedError,
+    UnknownAsset,
+    UnsupportedAsset,
+)
+from rotkehlchen.exchanges.data_structures import AssetMovement, MarginPosition, Trade
+from rotkehlchen.exchanges.exchange import ExchangeInterface, ExchangeQueryBalances
+from rotkehlchen.fval import FVal
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import (
+    deserialize_asset_amount,
+    deserialize_fee,
+    deserialize_price,
+    deserialize_timestamp,
+)
+from rotkehlchen.typing import (
+    ApiKey,
+    ApiSecret,
+    AssetMovementCategory,
+    Location,
+    Timestamp,
+    TradePair,
+    TradeType,
+)
+from rotkehlchen.user_messages import MessagesAggregator
+from rotkehlchen.utils.interfaces import cache_response_timewise, protect_with_lock
+from rotkehlchen.utils.misc import ts_now_in_ms
+from rotkehlchen.utils.serialization import rlk_jsonloads_dict
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+API_SYSTEM_CLOCK_NOT_SYNCED_ERROR_CODE = 400002
+# More understandable explanation for API key-related errors than the default `reason`
+API_KEY_ERROR_CODE_ACTION = {
+    400003: 'Invalid API key value.',
+    400004: 'Invalid API passphrase.',
+    400005: 'Invalid API secret.',
+    400007: 'Provided KuCoin API key needs to have "General" permission activated.',
+    411100: 'Contact KuCoin support to unfreeze your account',
+}
+API_PAGE_SIZE_LIMIT = 500
+# Rate limit is 1800 requests per minute, exceed it multiple times the system
```

Comment:
The documentations doesn't mention the related error code but the response status code (403 or 429). Neither their official client nor cctx use a particular code.

1800 per minutes is quite a big number so let me know whether it makes sense to reduce the `API_REQUEST_RETRIES_AFTER_SECONDS` (first time is 1s, then 2s - 1st retry, then 4s - 2nd retry).